nexus/src/db/datastore/mod.rs

```diff
@@ -103,14 +106,24 @@ impl<U, T> RunnableQuery<U> for T where
 
 pub struct DataStore {
     pool: Arc<Pool>,
+    resource_usage_producer: resource_usage::Producer,
```

Comment:
This struck me as odd at first -- why is _this_ the only piece of functionality in Nexus that hangs off the DataStore?  Having read more of this change it makes me wonder if all this belongs at the Nexus-layer, outside the DataStore?  The accounting seems like an application-layer concern.

Reply:
I'm not sure I understand why it would be a better fit moved up to the "Nexus" object.

Honestly, if datastore was split into more fine-grained classes, I would keep this as close to the `resource_usage` portion of it as possible - it's basically just a small cache of samples which get extracted by Clickhouse.

By keeping it closely coupled with the datastore, it is impossible for the application layer to "forget" to emit metrics when creating resources - when a `resource_usage` object is created, all "zeroed" samples are emitted immediately. When any individual resource is updated (e.g., virtual disk usage) we immediately add an appropriate sample to the producer.

I empathize with wanting to keep the datastore focused, but I also want to provide APIs that properly encapsulate behavior in a way that "doing the wrong thing" becomes difficult.

Reply:
To put the question differently: why is "resource reservation metrics" part of the datastore, but other application-level invariants (like [that every new Project gets a "default" VPC](https://github.com/oxidecomputer/omicron/blob/ae919f634fc25a667fb968592072267056b6d066/nexus/src/app/project.rs#L27-L75)) not?  Or: what criteria determine whether behavior belongs in the DataStore or application?

My mental model is: where the database can enforce a constraint (like uniqueness, or by using types), then it definitely belongs in the datastore.  After that, it gets a little hazy.  I see the appeal of well-enforced abstractions at the datastore layer, but I think one could follow that advice to move the whole application layer into the datastore, which eliminates the value of layering.

Reply:
This is an awkward fit for existing at the application layer, mostly because of the call stack. I *can* put it there, but then we'll have an even-more-tightly coupled relationship between the datastore and the application layer.

For example, suppose we're allocating an instance - our call stack looks roughly like the following

- (app layer) nexus/src/app/instance.rs:  `project_create_instance`, which makes a saga
- (app layer) nexus/src/app/sagas/instance.rs: Actual instance creation saga
- (db layer) nexus/db/datastore/instance.rs The saga calls into the datastore to actually provision instance records
- (db layer) nexus/db/datastore/virtual_provisioning_collection.rs The saga calls into the datastore to actually mark resources as provisioned.

Now, at this point, the call stack is strictly "app above db" - the application layer calls into the DB layer, not the other way around.

When we add metrics collection, we can do the following:

1. Add it directly to the "db" - this is what the PR currently does.
2. Add it at the "app" layer, call into it from `nexus/db/datastore/virtual_provisioning_collection.rs`. This kinda sucks, because it makes the DB layer "upcall" to the application layer, which makes them circularly dependent.
3. Add it at the "app" layer, but only update metrics when we return in the callstack back to  the application layer. This also kinda sucks, because it requires propagating all the metrics from pretty far downstack to pretty far upstack. IMO this is doable, but kinda gross, because it requires returning a lot of values, and ensuring that the caller remembers to actually hand them off to oximeter.

If you feel strongly about this, I can implement option (3) - this would involve taking all the calls to append to the `Producer` in `nexus/src/db/datastore/virtual_provisioning_collection.rs`, and instead invoking them further up the stack.